### PR TITLE
Add a walker loop

### LINKED ISSUE #97: Add a walker loop in drivers
Having a walker loop decoupled from threads is still very useful.
Need to cover all the drivers.

### PR DESCRIPTION
**Pull request by ye-luo**
Now the number of walkers can be control separately from the number of threads.
option [-w number_of_walkers]

Add a basic unit called mover which holds all the data.

Close #97

### CODE REVIEW

**Comment by mdewing** [comment]
The Mover class could use some documentation.

Also, what is the difference between a Mover and Walker?   Was it that the Mover contains temporary storage, and the Walker contains the minimal necessary state?

**Comment by ye-luo** [comment]
I will add docs. Correct, mover includes scratch space while walker contains minimal data in principle. A mover can handle walkers sequentially.

**Comment by mdewing** [comment]
As mentioned,  I think it would useful to have the documentation say what the scratch space is for (distance tables, inverse matrix, etc), and the size of each space.   I think it would be better all in one place here,  rather than scattered among the different wavefunction components.

**Comment by ye-luo** [comment]
The scratch space is held within each member objects. This won't change and there is no intention to use a global scratch space.
For example, a lightweight walker only carries particle coordinates but Mover.ParticleSet carries distance tables (consider this is a scratch space).

**Comment by mdewing** [comment]
My second sentence wasn't clear -  the documentation of the scratch space should be here, rather than make the user go to the documentation for each wavefunction component.

**Comment by ye-luo** [comment]
I still think if a wavefunction component uses some scratch space, the doc needs to be local to it. Maybe I didn't explain well what scratch space is. I mean every memory not provided by the lightweight walker but used during walking is scratch space.

**Review by prckent — CHANGES_REQUESTED** [review]
At least mover.hpp needs a substantial number of doxygen comments added:

- Explain the purpose of the class
- Label *all* the variables (rng, ions, els, spo, wavefunction, nlpp)
- Label *all* the functions. What is the purpose, expectations, outputs etc.?
- Generally try to make more accessible to someone reading the code for the first time (or Nth time)

File: src/Drivers/Mover.hpp [hunk 1]
```diff
@@ -27,28 +33,48 @@
...
   };
 
+  // Ye: the following two routines are used for initialize data.
+  //     Needs to be moved to more appropriate places like Input/
+  /// build the ParticleSet of ions
   int build_ions(ParticleSet &ions,
                  const Tensor<int, 3> &tmat,
                  Tensor<QMCTraits::RealType, 3> &lattice);
 
+  /// build the ParticleSet of ions
```
Comment (prckent) [inline src/Drivers/Mover.hpp]: Electrons?

Comment (ye-luo) [reply src/Drivers/Mover.hpp]: You are probably reviewing some old commits. Just refresh your browser.

**Review by prckent — APPROVED** [review]

**Comment by ye-luo** [comment]
@prckent Thanks.

**Pull request merged**